# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/arm64/MacroAssembler-arm64.h
# Commit: 28feda14fdec
# Full Hash: 28feda14fdec962ea68c26a0570ea69042a5ad27
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2023-08-05 09:19:01
# Regressor Bug: 1692065
# File Overlap Count: 3
# Description:
#   Bug 1692065 - wasm: Add tagging for JSString* in AnyRef. r=yury
#   
#   Before this commit we would need to box a JSString* in WasmValueBox
#   when we received one from JS and converted to an AnyRef.
#   
# ==============================================================================

diff -r 85605041672d -r 28feda14fdec js/src/jit/arm64/MacroAssembler-arm64.h
--- a/js/src/jit/arm64/MacroAssembler-arm64.h	Fri Aug 04 14:06:50 2023 +0000
+++ b/js/src/jit/arm64/MacroAssembler-arm64.h	Fri Aug 04 14:06:50 2023 +0000
@@ -1449,7 +1449,17 @@
         Operand(JS::detail::ValueGCThingPayloadMask));
   }
 
+  void unboxWasmAnyRefGCThingForGCBarrier(const Address& src, Register dest) {
+    loadPtr(src, dest);
+    And(ARMRegister(dest, 64), ARMRegister(dest, 64),
+        Operand(wasm::AnyRef::GCThingMask));
+  }
+
   // Like unboxGCThingForGCBarrier, but loads the GC thing's chunk base.
+  void getGCThingValueChunk(Register src, Register dest) {
+    And(ARMRegister(src, 64), ARMRegister(dest, 64),
+        Operand(JS::detail::ValueGCThingPayloadChunkMask));
+  }
   void getGCThingValueChunk(const Address& src, Register dest) {
     loadPtr(src, dest);
     And(ARMRegister(dest, 64), ARMRegister(dest, 64),
@@ -1460,6 +1470,11 @@
         Operand(JS::detail::ValueGCThingPayloadChunkMask));
   }
 
+  void getWasmAnyRefGCThingChunk(Register src, Register dest) {
+    And(ARMRegister(dest, 64), ARMRegister(src, 64),
+        Operand(wasm::AnyRef::GCThingChunkMask));
+  }
+
   inline void unboxValue(const ValueOperand& src, AnyRegister dest,
                          JSValueType type);
 